node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,158 @@
 package cni
 
 import (
+	"context"
 	"encoding/base64"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
+	"sync"
 	"time"
 
 	"github.com/sirupsen/logrus"
-	"gopkg.in/fsnotify/fsnotify.v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 )
 
-var serviceaccountDirectory string = "/var/run/secrets/kubernetes.io/serviceaccount/"
+const defaultCNITokenValiditySeconds = 3600
+const minTokenRetryDuration = 5 * time.Second
+const defaultRefreshFraction = 3
+
 var kubeconfigPath string = "/host/etc/cni/net.d/calico-kubeconfig"
 
-func Run() {
-	// Log to stdout.  this prevents our logs from being interpreted as errors by, for example,
-	// fluentd's default configuration.
-	logrus.SetOutput(os.Stdout)
+type TokenRefresher struct {
+	tokenSupported bool
+	tokenOnce      *sync.Once
+
+	namespace          string
+	serviceAccountName string
 
-	// Create a watcher for file changes.
-	watcher, err := fsnotify.NewWatcher()
+	tokenChan chan TokenUpdate
+}
+
+type TokenUpdate struct {
+	Token          string
+	ExpirationTime time.Time
+	Error          error
+}
+
+func NewTokenRefresher(namespace string, serviceAccountName string) *TokenRefresher {
+	return &TokenRefresher{
+		tokenSupported:     false,
+		tokenOnce:          &sync.Once{},
+		namespace:          namespace,
+		serviceAccountName: serviceAccountName,
+		tokenChan:          make(chan TokenUpdate),
+	}
+}
+
+func (t *TokenRefresher) UpdateToken() TokenUpdate {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		logrus.WithError(err).Error("Error generating kube config.")
+		return tokenUpdateError(err)
+	}
+	err = rest.LoadTLSFiles(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Error loading TLS files.")
+		return tokenUpdateError(err)
+	}
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to create clientset for CNI kubeconfig")
+		return tokenUpdateError(err)
+	}
+
+	validity := int64(defaultCNITokenValiditySeconds)
+	tr := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{"kubernetes"},
+			ExpirationSeconds: &validity,
+		},
+	}
+
+	tokenRequest, err := clientset.CoreV1().ServiceAccounts(t.namespace).CreateToken(context.TODO(), t.serviceAccountName, tr, metav1.CreateOptions{})
+	if apierrors.IsNotFound(err) && !t.tokenRequestSupported(clientset) {
```

Comment:
In the scenario that the token request API isn't supported, we should probably fall back on our old behavior of using the calico/node token. Otherwise I think there's a chance we break some users running on clusters without this API enabled. WDYT?

Reply:
I added the fall back case, which is a bit ugly as I needed to figure out the token validity period, but was not aware that a corresponding jwt library is used in calico already.
Looking at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/token/token_manager.go, it seems `kubelet` does this check whether the api is supported, but does not handle it at all, i.e. if the API is not supported it will be retried anyway. Therefore, I am not even sure if it is necessary to check for the API support or if it can be assumed to be always available.
Without the fallback code, the logic is certainly a bit easier to understand.

Reply:
Yeah, that is a bit curious that kubelet just retries anyway. 

The TokenRequest API does rely on certain arguments being passed to the kube-apiserver IIUC (although I suspect for the vast majority of clusters these arguments are present).

I'd rather be safe here and include a fallback than not, though.